src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/EnumConverterTests.cs

```diff
@@ -554,5 +555,63 @@ public enum SampleEnumSByte : sbyte
             F = 1 << 5,
             G = 1 << 6,
         }
+
+        [Fact]
+        public static void Honor_EnumNamingPolicy_On_Deserialization()
+        {
+            JsonSerializerOptions options = new()
+            {
+                Converters =
+                {
+                    new JsonStringEnumConverter(namingPolicy: new SimpleSnakeCasePolicy() )
+                }
+            };
+
+            BindingFlags bindingFlags = JsonSerializer.Deserialize<BindingFlags>(@"""non_public""", options);
+            Assert.Equal(BindingFlags.NonPublic, bindingFlags);
+
+            // Flags supported without naming policy.
+            bindingFlags = JsonSerializer.Deserialize<BindingFlags>(@"""NonPublic, Public""", options);
+            Assert.Equal(BindingFlags.NonPublic | BindingFlags.Public, bindingFlags);
+
+            // Flags supported with naming policy.
+            bindingFlags = JsonSerializer.Deserialize<BindingFlags>(@"""static, public""", options);
+            Assert.Equal(BindingFlags.Static | BindingFlags.Public, bindingFlags);
+
+            // Null not supported.
+            Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<BindingFlags>("null", options));
+
+            // Null supported for nullable enum.
+            Assert.Null(JsonSerializer.Deserialize<BindingFlags?>("null", options));
+        }
+
+        [Fact]
+        public static void EnumDictionaryKeyDeserialization()
+        {
+            JsonNamingPolicy snakeCasePolicy = new SimpleSnakeCasePolicy();
+            JsonSerializerOptions options = new()
+            {
+                Converters =
+                {
+                    new JsonStringEnumConverter(namingPolicy: snakeCasePolicy)
+                },
+                DictionaryKeyPolicy = snakeCasePolicy
+            };
+
+            // Baseline.
+            var dict = JsonSerializer.Deserialize<Dictionary<BindingFlags, int>>(@"{""NonPublic, Public"": 1}", options);
+            Assert.Equal(1, dict[BindingFlags.NonPublic | BindingFlags.Public]);
+
+            // DictionaryKeyPolicy not honored for dict key deserialization.
+            Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<Dictionary<BindingFlags, int>>(@"{""NonPublic0, Public0"": 1}", options));
+
+            // EnumConverter naming policy not honored.
+            Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<Dictionary<BindingFlags, int>>(@"{""non_public, static"": 0, ""NonPublic, Public"": 1}", options));
```

Comment:
Why though? this behavior seems surprising and is inconsisten with how we serialize. Values should be roundtrippable.

Reply:
WRT dictionary key policy and deserialization - https://github.com/dotnet/runtime/issues/30142. I think any policy changes here should be discussed in a JSON design forum.

The `EnumConverter` policy is used for enum "values" (collection elements, root enums, property values etc). I consider `options.PropertyNamingPolicy` analogous - it is not applied to dictionary keys. I believe the same should continue to apply for enum keys where we don't apply the `EnumConverter` policy.

Reply:
> WRT dictionary key policy and deserialization - #30142. I think any policy changes here should be discussed in a JSON design forum.

Let's continue that conversation in https://github.com/dotnet/runtime/pull/73348#discussion_r938711465

> The EnumConverter policy is used for enum "values" (collection elements, root enums, property values etc). I consider options.PropertyNamingPolicy analogous - it is not applied to dictionary keys

This is not true when it comes to serialization; the existing `WriteAsPropertyNameCore` implementation consults the `DictionaryKeyPolicy` and then the converter naming policy, in that order of precedence. This means that JSON produced by a certain `JsonSerializerOptions` options is explicitly failing to deserialize with that same configuration. Bare minimum we should strive to be consistent.

Reply:
> This is not true when it comes to serialization; the existing WriteAsPropertyNameCore implementation consults the DictionaryKeyPolicy and then the converter naming policy, in that order of precedence.

I've reverted some logic to keep this as-is (and avoid a breaking change).